crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs

```diff
@@ -0,0 +1,186 @@
+use ide_db::assists::{AssistId, AssistKind};
+use syntax::ast::{self, HasGenericParams, HasName};
+use syntax::{AstNode, SyntaxKind};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_nested_function_to_closure
+//
+// Converts a function that is defined within the body of another function into a closure.
+//
+// ```
+// fn main() {
+//     fn fo$0o(label: &str, number: u64) {
+//         println!("{}: {}", label, number);
+//     }
+//
+//     foo("Bar", 100);
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let foo = |label: &str, number: u64| {
+//         println!("{}: {}", label, number);
+//     };
+//
+//     foo("Bar", 100);
+// }
+// ```
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::Name>()?;
+    let function = name.syntax().parent().and_then(ast::Fn::cast)?;
+
+    if !is_nested_function(&function) || is_generic(&function) {
+        return None;
+    }
```

Comment:
Very much likely to be rare occasion, but I suppose we don't want to trigger this assist when the function has modifiers i.e. `async`, `const`, and/or `unsafe`. For unsafe, We could make the closure body an unsafe block though.

Reply:
> For unsafe, We could make the closure body an unsafe block though.

That doesn't have the same semantics though, so I would say not to. (Calling the closure would be safe, opposed to the function which would be unsafe to call)

Reply:
Right! I have no idea why I thought that was okay :/